Fix dumping issues with / and the first sub level

Kidswiss opened the pull request:
There was an issue that prevented the dump command from working
correctly when either:

* `/` contained multiple nodes (e.g. `restic backup /`)
* dumping a file in the first sublevel was attempted (e.g. `/foo`)



<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

What is the purpose of this change? What does it change?
--------------------------------------------------------
Fixes #2254
<!--
Describe the changes here, as detailed as needed.
-->


<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "closes #1234" so that the issue is
closed automatically when this PR is merged.
-->

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

Kidswiss committed:
dump: Fix dumping issues with / and the first sub level

There was an issue that prevented the dump command from working
correctly when either:

* `/` contained multiple nodes (e.g. `restic backup /`)
* dumping a file in the first sublevel was attempted (e.g. `/foo`)

fd0 commented:
Hm. What about moving the tar code into a package below `internal/` and adding the cases you tested manually as real tests? I can give it a shot if you like me to help :)

Kidswiss commented:
That's a great idea, especially the test cases. Do you have anything to add to the ones I already declared at #2254?

I'd appreciate a few hints on what you have in mind, so we're on the same level. Then I'll give it a shot :)

fd0 commented:
So, what I'd do (roughly):
 * Remove all `tar` code from `printFromTree()`
 * Move the `tar` code to e.g. `internal/archive`
 * Export a function in `internal/archive`, something along the lines `func WriteTar(ctx context.Context, repo restic.Repository, tree restic.ID, dst io.Writer) error`
 * Write a function `isDir()` (or so) to decide if the target is a dir or a file
 * then either call `printFromTree()` or `archive.WriteTar()`
 * For testing, add tests in the `internal/archive` package on a test repo (look at the archiver tests if you need inspiration):
   * for each test, create a repo with a known structure, run `WriteTar()` on it writing the output to a `bytes.Buffer`
   * then read the output back with a helper function that checks the `tar` file against the expected structure

It's a big change now thinking about it, so please just let me know if you need help with anything!

It's important to split the change into as many small commits as necessary, this makes reviewing so much easier!

fd0 commented on cmd/restic/cmd_dump.go:
```diff
@@ -90,13 +90,31 @@ func printFromTree(ctx context.Context, tree *restic.Tree, repo restic.Repositor
+
+		// If we print / we need to asume that there are multiple nodes at that
+		// level in the tree.
+		if pathToPrint == "/" {
```
I'd handle this special case first in the function (after the sanity checks), so we get it out of the way.

fd0 commented on cmd/restic/cmd_dump.go:
```diff
@@ -90,13 +90,31 @@ func printFromTree(ctx context.Context, tree *restic.Tree, repo restic.Repositor
+		// level in the tree.
+		if pathToPrint == "/" {
+			tw := tar.NewWriter(os.Stdout)
+			defer tw.Close()
```
`Close()` returns an error which isn't checked here, but needs to be. I'd remove this line and handle the case explicitly as shown below. I'd also not use `defer()`, because error checking in `defer` is very error prone.
```suggestion
```

fd0 commented on cmd/restic/cmd_dump.go:
```diff
@@ -90,13 +90,31 @@ func printFromTree(ctx context.Context, tree *restic.Tree, repo restic.Repositor
+				rootNode.Path = pathToPrint
+				err := tarTree(ctx, repo, rootNode, pathToPrint, tw)
+				if err != nil {
+					return err
```
```suggestion
					_ = tw.Close()
					return err
```

fd0 commented on cmd/restic/cmd_dump.go:
```diff
@@ -90,13 +90,31 @@ func printFromTree(ctx context.Context, tree *restic.Tree, repo restic.Repositor
+					return err
+				}
+			}
+			return nil
```
```suggestion
			return tw.Close()
```

fd0 commented on cmd/restic/cmd_dump.go:
```diff
@@ -90,13 +90,31 @@ func printFromTree(ctx context.Context, tree *restic.Tree, repo restic.Repositor
 				node.Path = pathToPrint
-				return tarTree(ctx, repo, node, pathToPrint)
+				tw := tar.NewWriter(os.Stdout)
+				defer tw.Close()
```
You're dropping an error returned by `tw.Close()`, which needs to be checked. I'd remove this line and handle it as shown below.
```suggestion
```

fd0 commented on cmd/restic/cmd_dump.go:
```diff
@@ -90,13 +90,31 @@ func printFromTree(ctx context.Context, tree *restic.Tree, repo restic.Repositor
-				return tarTree(ctx, repo, node, pathToPrint)
+				tw := tar.NewWriter(os.Stdout)
+				defer tw.Close()
+				return tarTree(ctx, repo, node, pathToPrint, tw)
```
```suggestion
				err = tarTree(ctx, repo, node, pathToPrint, tw)
				if err != nil {
					_ = tw.Close()
					return err
				}
				return tw.Close()
```

fd0 commented on cmd/restic/cmd_dump.go:
```diff
@@ -194,28 +212,28 @@ func getNodeData(ctx context.Context, output io.Writer, repo restic.Repository,
 	}
 
+	// If this is a file we are finished
+	if rootNode.Subtree == nil {
```
That's not the correct way to test for a tree, use this:
```suggestion
	if rootNode.Type != "tree" {
```

Kidswiss replied:
Hmmm... the goal is rather to test for a file.

So I guess:
```
if rootNode.Type == "file"
```

Would be the best solution here? As your suggestion would also return if the node is a folder or a link.

fd0 commented on cmd/restic/cmd_dump.go:
```diff
@@ -54,12 +54,9 @@ func init() {
-	if d == "/" {
-		return []string{d}
-	}
 	s := splitPath(path.Clean(d))
```
There's some weird stuff going on when you pass odd paths, like `restic dump latest /foo/../`, which is caused by `path.Clean()` returning either an empty string or a relative path that makes no sense. You could do the following to always have it be absolute:
```suggestion
	s := splitPath(path.Clean(path.Join("/", d)))
```

fd0 reviewed (dismissed):
Thanks a lot for tackling these issues!

I had a look and found a couple of odd things, in addition to the line comments:
 * The path in the `tar` varies whether or not the path given to `dump` is absolute:
   * running `restic dump latest /cmd` gives me paths like `/cmd/restic/acl.go` in the tar
   * running `restic dump latest cmd` gives me paths like `cmd/restic/acl.go` in the tar
   * I think the paths in the resulting `tar` file should be relative by default, what are your thoughts on that?
 * The help text for the `dump` command isn't up to date, it does not mention writing `tar` files at all

fd0 commented:
As always, please let me know if you need help, we can also resolve this collaboratively if you want.

Kidswiss commented:
Hey, thanks for the feedback!

I may have some time tomorrow that I can implement your suggestions and fix the odd stuff. So I can do the refactoring on a clean base.

This feature has quite some hidden caveats that make it more complex than anticipated, but we're getting there! :)

I like the idea of possible support for other formats, too. Also just wondering what your opinion of #2226 is. As decisions we make here could have potential impact on the implementation of the feature.

fd0 commented:
I've taken the liberty of rebasing on master, so that the CI tests on Windows pass.

Kidswiss commented:
Sorry for the radio silence, I was on a holiday :)

I should be able to do some work here this week. But as there's much to do at work I'll have to do it during my free time. So don't expect huge leaps in progress :)

Commit:
dump: Extract tar code to internal/dump

Kidswiss commented:
I moved everything related to `tar` to the archiver. The whole tar handling is now much more consistent and should work better.

Although I'll need to test it on a windows machine, too. Just to be sure if the original issue is now fixed.

Todo:
* implement tests

Commit:
dump: Always dump relative paths into tarballs

Tarballs should only contain relative paths.

Kidswiss commented:
Hi @fd0 

I had a look at the test code in Archiver and I'm not quite sure what exactly I'll need, to create a test repo. I feel like I need something like `TestArchiverSaveDir` to create the repo. But as I'm not sure I'd rather ask you before I start implementing that and realise that it's not the right way todo that :)

Commit:
dump: Added unit tests for tar output

Kidswiss commented:
I added some automated test for various cases where `/` is dumped. I compare the created tar file against the temporary folder created by the restic tests and if I find some differences the test fails.

Kidswiss commented:
@fd0 any chance you could have a look at this so I can finish this up cleanly? :) Still not quite sure my test code is really doing the right thing...

Kidswiss commented:
I just realised that the push on 9th August was something completely broken and basically reverted everything... I fixed that now.

Doing this I found that the CI build for MacOS fails with the new tar tests I implemented, due to timestamps not matching. This is interesting as I'm developing locally with MacOS and can't reproduce this.

Commit:
Add changelog

Kidswiss commented:
@fd0 @rawtaz 

Finally got around to clean this up properly. Any chance one of you guys have time for a review? I'd really like to get this into the next release. Thanks! :)

greatroar commented:
Why is tar dumping implemented in internal/archiver? That package is for backing up, not for accessing backups. Wouldn't internal/dump make more sense?

Kidswiss commented:
@greatroar Thanks for the feedback!

It's quite a while ago since this change was started. But it was @fd0's suggestion here:  https://github.com/restic/restic/pull/2255#issuecomment-487309292. Not quite sure about his reasoning though. Maybe because `tar` is an archive format? It seemed reasonable at the time.

The actual dump logic is in the `cmd_dump.go` file. The logic in the `internal/archive` packe is solely used to create the tar archives given some arbitrary readers.

MichaelEischer committed:
dump: Add test for splitPath

MichaelEischer reviewed (commented):
Sorry for us taking so long to get back to you. 

I've rebased this PR to the current master and restructured the commits and their commit messages a bit (but I've checked that nothing relevant did change).

Besides that the tar code now lives in `internal/dump`. @fd0's suggestion was `internal/archive` (not archiver), but that's probably quite easy to misread. This required some changes to the test as it is no longer able to access the internal of the archiver.

I've also removed the redundant call to `path.Clean` in `splitPath` (`path.Join` already cleans the path). 

I'll try whether I can get the tar test to fully work on windows and the modtime check without introducing too much complexity.

I nearly forgot to mention: I'll have to test this PR a bit more, but besides that it should be ready for merging.

Kidswiss commented:
Hi @MichaelEischer 

Thanks for the help with this Pull Request. 

Ah yes, I completely miss understood @fd0's suggestions then, sorry about that.

Kidswiss commented on changelog/unreleased/pull-2255:
```diff
@@ -0,0 +1,10 @@
+Bugfix: Fix tar issues when dumping `/`
+
+We've fixed and issue with dumping `/` to tar. As well as dumping files on the
+first sublevel e.g. `/foo`. This also fixes tar dumping issues on Windows where
```
It's quite a while ago that I had a look at this for Windows.

The issue for Windows was basically the same as when you're dumping `/` on Linux. Where `/` under linux contains multiple nodes that weren't iterated correctly in the first version of the tar dumping code.

But I'm not really sure any more what exact circumstances caused this problem on Windows unfortunately, as that was over a year ago.

MichaelEischer replied:
@Kidswiss Is there a special reason to mention windows explicitly here? The tar dumping just seems to have been as broken as on other systems?

Kidswiss replied:
No problem :)

If you're happy with the changelog text we can resolve this discussion.

MichaelEischer commented:
I think I've found the reason for the test failures: the tar header contains the timestamps rounded to the next second. `os.Stat` on the other hand provides sub-second precision. This meant that the "last modified"-timestamps were different in general. Together with a missing negation of the equality check this means that the test only succeeded if both timestamps were different, which usually was the case.

The timestamp from `os.Stat` is now also rounded to seconds and thus should correctly match the one from the tar header.

My manual tests with dumping a tar file restricted to root or the first sub-level were also successfully.

MichaelEischer committed:
dump: Properly handle file paths without / prefix

filepath.Rel failed if the requested path did not start with a `/` e.g.
`restic` instead of `/restic`.

MichaelEischer committed:
dump: Use slashes as path separators in tar on windows

This conversion is necessary as paths returned by filepath.Rel use the
operating-system native path separator

MichaelEischer committed:
dump: Mention tar output in command help text

MichaelEischer commented:
I've fixed a few further small issues and updated the command description.

@Kidswiss Could you take a quick look at the three newest commits?

I've also done a bit of testing on windows, the generated tar files seem mostly fine. But I've noticed that 7zip failed to open the tar archives once they contained directories. But that is caused by #2319 and should by fixed in another PR.

Kidswiss commented:
Just had a look at the commits. Looks good to me!

Thanks for the Windows testing. I agree that the issue should be handled in #2319.

MichaelEischer reviewed (approved):
LGTM
